AddressSpace/src/FreeVariablesEngine.cpp

```diff
@@ -0,0 +1,209 @@
+    if (convertedNumber < rangeMin || convertedNumber > rangeMax)
+        throw_runtime_error_with_origin("value '" + str + "' out of allowed range '" + std::to_string(rangeMin) + "," + std::to_string(rangeMax) + "'");
+    return convertedNumber;
+
+}
+
+//! The version for integer does range check, the version for floating points does not do it.
+template<typename T, typename Twider> // T here will be e.g. OpcUa_UInt16 and Twider wider will be uint32_t for which a standard conversion function exists
+static void initializeIntegerNumeric (
+        UaVariant& out,
+        void (UaVariant::*setter)(T),
+        Twider(*convFunc)(const std::string&, std::size_t*, int base),
+        const Configuration::FreeVariable::initialValue_optional& configInitialValue)
+{
+    T initialValue (0);
+    if (configInitialValue.present())
+        initialValue = veryIntelligentWrapperOfConversionFunctionWithRangeCheck<T, Twider>( configInitialValue.get(), convFunc );
+    (out.*setter)(initialValue);
+}
+
+template<typename T> // T here will be e.g. OpcUa_Float or OpcUa_Double
+static void initializeFloatingPointNumeric (
+        UaVariant& out,
+        void (UaVariant::*setter)(T),
+        T(*convFunc)(const std::string&, std::size_t*),
+        const Configuration::FreeVariable::initialValue_optional& configInitialValue)
+{
+    T initialValue (0);
+    if (configInitialValue.present())
+        initialValue = veryIntelligentWrapperOfConversionFunction( configInitialValue.get(), convFunc );
+    (out.*setter)(initialValue);
+}
+
+void FreeVariablesEngine::instantiateFreeVariable(
+    AddressSpace::ASNodeManager* nm,
+    UaNodeId                     parentNodeId,
+    const                        Configuration::FreeVariable& config
+    )
+{
+    AddressSpace::ChangeNotifyingVariable* freeVariable = new AddressSpace::ChangeNotifyingVariable (
+      nm->makeChildNodeId(
+          parentNodeId,
+          config.name().c_str()),
+          config.name().c_str(),
+      nm->getNameSpaceIndex(),
+      UaVariant(),
+      OpcUa_AccessLevels_CurrentReadOrWrite,
+      nm);
+
+    // data type
+    freeVariable->setDataType(freeVariableTypeToBuiltInType(config.type()));
+
+    // type-dependent business, the order is as per Oracle.AllQuasarDataTypes
+    UaVariant initialValue;
+    if (config.type() == "Boolean")
+    {
+        OpcUa_Boolean value = OpcUa_False; // default if not present
+        if (config.initialValue().present())
+        {
+            std::string initialValueAsString = config.initialValue().get();
+            if (initialValueAsString != "1" && initialValueAsString != "0") // booleans are from discrete set.
+            {
+                throw_runtime_error_with_origin("While instantiating FreeVariable '"
+                        + config.name() + "' under '"
+                        + parentNodeId.toString().toUtf8()
+                        + " boolean literal is wrong '" + initialValueAsString + "'. Allowed values are 0 or 1.");
```

Comment:
why not also true or false?

Reply:
I'd prefer to stick to one version so either "true"/"false" or "1"/"0". "true"/"false" seems more correct to me, so I'll rework to that.